Add tutorials for using the training script and (#196)
huggingface/lerobot: examples/4_train_policy_with_script.md

> This tutorial will explain the training script, how to use it, and particularly the use of Hydra to configure everything needed for the training run.

Reviewer: This in-depth tutorial is important, but it would be nice to have a specific markdown file called: `5_reproduce_sota.md` with a structured list of training commands + pointers to our model page on the hub + evaluation commands of the pretrained models.

```
# Table of results

TODO Table with link to model page on the hub + link to section in the markdown file

## Diffusion policy on Pusht

Model card on the hub:

Training command:
```
TODO
```

Eval Command
```
TODO
```

## Aloha

...

```

Author: I don't think this is a good idea. It adds more not-maintainable documentation. These details live with the model cards and I don't think they should be repeated elsewhere. IMO, we should just have a zoo.md style file that points to the model cards.

Reviewer: What do you think of adding a `5_reproduce_sota.md` which would be a tutorial on how to access the model card list, and how from the model card, access the training command?

Maybe long term with can avoid this maintainable documentation, but right now I have no idea how to access these training commands.

Author: How about I put the commands on the model cards. Then you have command, configuration, and commit hash all in one place. Then I can add a section in the main README to talk about the model cards.

Author: Added to this PR.

On your thought: I suppose... it's a one-liner right now, so a script doesn't add much value right?

Author: Ah for the README. Yeah that's a good idea :)

## The training script

LeRobot offers a training script at [`lerobot/scripts/train.py`](../../lerobot/scripts/train.py). At a high level it does the following:

- Loads a Hydra configuration file for the following steps (more on Hydra in a moment).
- Makes a simulation environment.
- Makes a dataset corresponding to that simulation environment.
- Makes a policy.
- Runs a standard training loop with forward pass, backward pass, optimization step, and occasional logging, evaluation (of the policy on the environment), and checkpointing.

## Our use of Hydra

Explaining the ins and outs of [Hydra](https://hydra.cc/docs/intro/) is beyond the scope of this document, but here we'll share the main points you need to know.

> First, consider that `lerobot/configs` might have a directory structure like this (this is the case at the time of writing):

Reviewer: ```suggestion
First, `lerobot/configs` has a directory structure like this:
```

Author: I'd like this to be more concrete as well. But being a little vague allows room for error as this is a very not-maintainable bit of documentation.

@aliberts do you think we can use a tool like this in CI? https://github.com/UmbrellaDocs/linkspector

Author: @Cadene I made my sentence less apologetic, leaving just one "might".

Reviewer: I dont think we should use "might". We should tell exactly what is expected. No?

Author: Did might -> will (everywhere), and used "like" or "something like". There may be minor refactors that have no bearing on the essentials that this document tries to convey, so I don't want the language to overly-commit.

```
.
├── default.yaml
├── env
│   ├── aloha.yaml
│   ├── pusht.yaml
│   └── xarm.yaml
└── policy
    ├── act.yaml
    ├── diffusion.yaml
    └── tdmpc.yaml
```

**_For brevity, in the rest of this document we'll drop the leading `lerobot/configs` path. So `default.yaml` really refers to `lerobot/configs/default.yaml`._**

> When you run the training script, Hydra takes over via the `@hydra.main` decorator. If you take a look at the `@hydra.main`'s arguments you will see `config_path="../configs", config_name="default"`. This means Hydra looks for `default.yaml` in `../configs` (which resolves to `lerobot/configs`).

Reviewer: I would suggest to first give one or more examples, before delving into this complex details.

Author: Hmm, I'm not sure what examples to add that would build the reader up to understanding this point any better. I've added a couple of lines that might help with train-of-thought. Let me know if you have a better approach.

Among regular configuration hyperparameters like `device: cuda`, `default.yaml` has a `defaults` section. It might look like this.

```yaml
defaults:
  - _self_
  - env: pusht
  - policy: diffusion
```

> So, Hydra will grab `env/pusht.yaml` and `policy/diffusion.yaml` and incorporate their configuration parameters (any configuration parameters already present in `default.yaml` are overriden).

Reviewer: Suggestion
"""
`default.yaml` is always the entry point. It begins by a `defaults` section likes this:

```yaml
defaults:
  - _self_
  - env: pusht
  - policy: diffusion
```

It means that Hydra will grab `env/pusht.yaml` and `policy/diffusion.yaml` and incorporate their configuration parameters (any configuration parameters already present in `default.yaml` are overriden).

After this `defaults` section, `default.yaml` also contains regular configuration hyperparameters like `device: cuda`, or .... TODO.
"""

Author: Done, with some further tweaks.

## Running the training script with our provided configurations

> If you want to train Diffusion Policy with PushT, you really only need to run:

Reviewer: Maybe put this just after explaining the `defaults` section to not interrupt the reading flow with a `<h2>` (`## Running the....`)

```suggestion
To reproduce Diffusion Policy results on PushT, you only need to run:
```

Author: I'd rather leave the focus on the topic at hand, and add as an aside that we've structured things such that our configs are designed to reproduce a given variation of the original work. Ptal.

```bash
python lerobot/scripts/train.py
```

That's because `default.yaml` already defaults to using Diffusion Policy and PushT. To be more explicit, you could also do the following (which would have the same effect):

```bash
python lerobot/scripts/train.py policy=diffusion env=pusht

> ```

Reviewer: Yes!

> If you want to train ACT with Aloha, you can do:

Reviewer: You can also mention the running time.

```suggestion
To reproduce ACT results on Aloha (TODO say the environement and task), run this:
```

Author: As I've mentioned in other comments here, I want to keep the focus on "how do I use LeRobot". There are other resources for the specific details of reproducing SOTA models.

Reviewer: I think there are only these use cases:
- reproduce SOTA results (+ tune some hyperparameters, which are a diff with respect to SOTA hyerparameters)
- train on your own dataset
- train your own model on Aloha dataset

I dont think people are interested about "train ACT with Aloha" per se.

Author: I added a comment at the start of the advanced tutorial to make it clear that it addresses your point 2. Point 1 is covered in this tutorial with the changes I've made. For point 3, this is really dependent on the model, but I think the advanced tutorial helps.

```bash
python lerobot/scripts/train.py \
    policy=act \
    dataset_repo_id=lerobot/aloha_sim_transfer_cube_human \
    env=aloha \
    env.task=AlohaTransferCube-v0
```

There's something new here. Notice the `.` delimiter used to traverse the configuration hierarchy.

Putting all that knowledge together, here's the command that was used to train https://huggingface.co/lerobot/act_aloha_sim_transfer_cube_human.

```bash
python lerobot/scripts/train.py \
    hydra.run.dir=outputs/train/act_aloha_sim_transfer_cube_human \
    device=cuda
    env=aloha \
    env.task=AlohaTransferCube-v0 \
    dataset_repo_id=lerobot/aloha_sim_transfer_cube_human \
    policy=act \
    training.eval_freq=10000 \
    training.log_freq=250 \
    training.offline_steps=100000 \
    training.save_model=true \
    training.save_freq=25000 \
    eval.n_episodes=50 \
    eval.batch_size=50 \
    wandb.enable=false \
```

There's one new thing here: `hydra.run.dir=outputs/train/act_aloha_sim_transfer_cube_human`, which specifies where to save the training output.

---

Now, why don't you try running:

```bash
python lerobot/scripts/train.py policy=act env=pusht dataset_repo_id=lerobot/pusht
```

> That was a little mean of us, because if you did try running that code, you almost certainly got an exception of sorts. That's because there are aspects of the ACT configuration that are specific to the ALOHA environments, and here we have tried to use PushT.

Reviewer: Not sure about this one ^^

Author: Please let me know which part you're referring to. Is it the informal tone in "that was a little mean of us"?

Reviewer: If we do reference a command that doesnt work, we shouldnt say "you almost certainly got an exception of sorts.".
Instead, we expect to get a specific error. 
But I really not sure we should in this context.

Reviewer 2: Maybe you could clarify that the policy config file contains values (dataset_repo_id) that are specific to the combination of policy and environment. If the environment config does not match what the policy configuration expects, an error is thrown.

Author: @Cadene I don't want to be explicit about the exception. This will almost certainly change, and IMO it adds no value to explain which one it is.

@dirkmcpherson I think that's a good suggestion.

Ptal

> This tutorial will explain the training script, how to use it, and particularly the use of Hydra to configure everything needed for the training run.

Author: I think we discussed in another PR that `4_calculate_validation_loss.py` should be under advanced. So I'm thinking we move it there (removing the 4_), and leave this here as it follows nicely from example 3. What do you think?

## Our use of Hydra

Explaining the ins and outs of [Hydra](https://hydra.cc/docs/intro/) is beyond the scope of this document, but here we'll share the main points you need to know.

First, `lerobot/configs` has a directory structure like this:

```
.
├── default.yaml
├── env
│   ├── aloha.yaml
│   ├── pusht.yaml
│   └── xarm.yaml
└── policy
    ├── act.yaml
    ├── diffusion.yaml
    └── tdmpc.yaml
```

**_For brevity, in the rest of this document we'll drop the leading `lerobot/configs` path. So `default.yaml` really refers to `lerobot/configs/default.yaml`._**

When you run the training script with

```python
python lerobot/scripts/train.py
```

Hydra takes over via the `@hydra.main` decorator. If you take a look at the `@hydra.main`'s arguments you will see `config_path="../configs", config_name="default"`. This means Hydra looks for `default.yaml` in `../configs` (which resolves to `lerobot/configs`).

Therefore, `default.yaml` is the first configuration file that Hydra considers. At the top of the file, is a `defaults` section which looks likes this:

```yaml
defaults:
  - _self_
  - env: pusht
  - policy: diffusion
```

> So, Hydra then grabs `env/pusht.yaml` and `policy/diffusion.yaml` and incorporates their configuration parameters as well (any configuration parameters already present in `default.yaml` are overriden).

Author: There is a default for dataset_repo_id in order to match the policy and env defaults (I'm assuming... I didn't actually set this up).

> So, Hydra then grabs `env/pusht.yaml` and `policy/diffusion.yaml` and incorporates their configuration parameters as well (any configuration parameters already present in `default.yaml` are overriden).

Author: I will take your suggestions including the override disclaimer.

> we want to avoid override as much as possible.

We can do this by removing a bunch of params in default.yaml. Let's make that a different PR though, this one's gone on long enough.

Below the `defaults` section, `default.yaml` also contains regular configuration parameters.

> If you want to train Diffusion Policy with PushT, you really only need to run:

Author: Done. SOTA results are mentioned below.

```bash
python lerobot/scripts/train.py
```

> That's because `default.yaml` already defaults to using Diffusion Policy and PushT. To be more explicit, you could also do the following (which would have the same effect):

Author: Done.

```bash
python lerobot/scripts/train.py policy=diffusion env=pusht
```

> If you want to train ACT with Aloha, you can do:

Author: Done

```bash
python lerobot/scripts/train.py policy=act env=aloha
```

> **Notice, how the config overrides are passed** as `param_name=param_value`. This is the format the Hydra excepts for parsing the overrides.

Author: I get your point, and I've incorporated it slightly differently.

_As an aside: we've set up our configurations so that they reproduce state-of-the-art results from papers in the literature._

## Overriding configuration parameters in the CLI

> If you look in `env/aloha.yaml` you will see something like:

Author: Done.

```yaml
# lerobot/configs/env/aloha.yaml
env:
  task: AlohaInsertion-v0
```

And if you look in `policy/act.yaml` you will see something like:

```yaml
# lerobot/configs/policy/act.yaml
dataset_repo_id: lerobot/aloha_sim_insertion_human
```

But our Aloha environment actually supports a cube transfer task as well. To train for this task, you _could_ modify the two configuration files respectively.

We need to select the cube transfer task for the ALOHA environment.

```yaml
# lerobot/configs/env/aloha.yaml
env:
   task: AlohaTransferCube-v0
```

> We also need to use the cube transfer dataset.

Author: Unless you feel strongly about it, I think this adds redundancy and clutter and prefer to leave it mostly as is. to make it clear, I've turned the snippet into a diff

```yaml
# lerobot/configs/policy/act.yaml
dataset_repo_id: lerobot/aloha_sim_transfer_cube_human
```

Now you'd be able to run:

```bash
python lerobot/scripts/train.py policy=act env=aloha
```

and you'd be training and evaluating on the cube transfer task.

> OR, your could leave the configuration files in their original state and override the defaults via the command line:

Author: I don't want to say this as, I do get them to manually modify the config in the advanced tutorial.

```bash
python lerobot/scripts/train.py \
    policy=act \
    dataset_repo_id=lerobot/aloha_sim_transfer_cube_human \
    env=aloha \
    env.task=AlohaTransferCube-v0
```

There's something new here. Notice the `.` delimiter used to traverse the configuration hierarchy.

> Putting all that knowledge together, here's the command that was used to train https://huggingface.co/lerobot/act_aloha_sim_transfer_cube_human.

Author: Unfortunately, `:` gets included in the hyperlink. Leaving as is.

```bash
python lerobot/scripts/train.py \
    hydra.run.dir=outputs/train/act_aloha_sim_transfer_cube_human \
    device=cuda
    env=aloha \
    env.task=AlohaTransferCube-v0 \
    dataset_repo_id=lerobot/aloha_sim_transfer_cube_human \
    policy=act \
    training.eval_freq=10000 \
    training.log_freq=250 \
    training.offline_steps=100000 \
    training.save_model=true \
    training.save_freq=25000 \
    eval.n_episodes=50 \
    eval.batch_size=50 \
    wandb.enable=false \
```

There's one new thing here: `hydra.run.dir=outputs/train/act_aloha_sim_transfer_cube_human`, which specifies where to save the training output.

> ---

Author: I'm not sure what you mean by "load from a config yaml file inside an experiment checkpoint". I do have a PR in the works that does checkpointing and training resumption, and I plan to add a section here on how to resume training.
